### ISSUE TITLE
submit conformance results

### ISSUE THREAD

**Issue by BenTheElder** [kubernetes-sigs/kind#1374]
We run lots of conformance CI but we've not been submitting results regularly, we're about to lose our certification actually!

We should:
- submit results
- plan to submit more regularly

/help

**Comment by BenTheElder** [comment]
this can't wait any longer
/assign
/remove-help

**Comment by BenTheElder** [comment]
hit https://github.com/vmware-tanzu/sonobuoy/issues/746 /  https://github.com/vmware-tanzu/sonobuoy/issues/1086#issuecomment-595330535 on v1.15

skipping to v1.16. it would be unfortunate to not certify v1.15, but using sonobuoy to do the official testing has been frustrating here vs just using e2e.test and I'd rather make sure we at least submit something newer in time.

**Comment by BenTheElder** [comment]
v1.16 is running... on kind v0.7.0. will check in later with results.

**Comment by BenTheElder** [comment]
https://github.com/cncf/k8s-conformance/pull/906 for 1.16

**Comment by BenTheElder** [comment]
for v1.15 I fell back on the upstream conformance image / sonobuoy image results https://github.com/cncf/k8s-conformance/pull/907 which do not actually use sonobuoy but are maintained in part by the sonobuoy maintainers anyhow.

**Comment by BenTheElder** [comment]
we've submitted 1.15 & 1.16 succesfully, still need to do 1.17 and 1.18

**Comment by stealthybox** [comment]
I was able to automate this a bit.
`1.18.0` passed first try so I've submitted it:
https://github.com/cncf/k8s-conformance/pull/939

I ran the full suite for `1.17.0` using the same `run.sh` script, and it failed on this one test twice:
```
failed tests: 1
[sig-api-machinery] Aggregator Should be able to support the 1.10 Sample API Server using the current Aggregator [Conformance]
```

**Comment by stealthybox** [comment]
<details>
<summary><b>run.sh</b> (click to expand)</summary>

```shell
#!/usr/bin/env sh

v_k8s="${1:-"1.18.0"}"

e2e_focus="${e2e_focus}"
rerun_failed="${rerun_failed}"

set -eu

folder="v${v_k8s}_$(date +"%F_%H:%M.%S")"

# https://github.com/kubernetes-sigs/kind/releases/tag/v0.7.0
image_tags="
v1.18.0@sha256:0e20578828edd939d25eb98496a685c76c98d54084932f76069f886ec315d694
v1.17.0@sha256:9512edae126da271b66b990b6fff768fbb7cd786c7d39e86bdf55906352fdf62
v1.16.4@sha256:b91a2c2317a000f3a783489dfb755064177dbc3a0b2f4147d50f04825d016f55
v1.15.7@sha256:e2df133f80ef633c53c0200114fce2ed5e1f6947477dbc83261a6a921169488d
v1.14.10@sha256:81ae5a3237c779efc4dda43cc81c696f88a194abcc4f8fa34f86cf674aa14977
v1.13.12@sha256:5e8ae1a4e39f3d151d420ef912e18368745a2ede6d20ea87506920cd947a7e3a
v1.12.10@sha256:68a6581f64b54994b824708286fafc37f1227b7b54cbb8865182ce1e036ed1cc
v1.11.10@sha256:e6f3dade95b7cb74081c5b9f3291aaaa6026a90a977e0b990778b6adc9ea6248
"

v_tag="$( echo "${image_tags}" | grep "${v_k8s}" )"

cat << EOF \
  | kind create cluster \
    --name "${v_k8s}-conformance" \
    --image="kindest/node:${v_tag}" \
    --config=/dev/stdin \
    || true
apiVersion: kind.x-k8s.io/v1alpha4
kind: Cluster
nodes:
- role: control-plane
- role: worker
- role: worker
EOF

ctx="--context=kind-${v_k8s}-conformance"

sonobuoy delete "${ctx}" --wait

if [ "${e2e_focus}" ]; then
  echo e2e_focus
  folder="${folder}-e2e_focus"
  sonobuoy run "${ctx}" --wait "--e2e-focus=${e2e_focus}"
elif [ "${rerun_failed}" ]; then
  echo rerun
  folder="${folder}-rerun"
  sonobuoy e2e "${ctx}" --wait "${rerun_failed}" --rerun-failed
else
  echo certified_conformance_run
  sonobuoy run "${ctx}" --wait --mode=certified-conformance
fi

outfile="$(sonobuoy retrieve "${ctx}")"
mkdir -p "./${folder}"
tar xzf "${outfile}" -C "./${folder}"

sonobuoy results --plugin=e2e "${outfile}"
```

</details>

The above script is the one from the conformance README.
We can use it to create new clusters or re-use existing ones to run tests .

I can consistently get this test to fail on `1.17.0` but succeed on `1.18.0`:
```shell
kind delete cluster --name 1.17.0-conformance
e2e_focus="Sample API Server" ./run.sh 1.17.0  # fails
e2e_focus="Sample API Server" ./run.sh 1.17.0  # fails
e2e_focus="Sample API Server" ./run.sh 1.17.0  # fails

kind delete cluster --name 1.18.0-conformance
e2e_focus="Sample API Server" ./run.sh 1.18.0  # passes
```

Running this 1 test from a new cluster takes about 5 minutes.
Reusing the cluster takes 1-2 minutes.

It's notable that the `1.17.0` test and `1.18.0` test descriptions differ:
```shell
# 1.17.0
failed tests:
[sig-api-machinery] Aggregator Should be able to support the 1.10 Sample API Server using the current Aggregator [Conformance]

# 1.18.0
passed tests: 1
[sig-api-machinery] Aggregator Should be able to support the 1.17 Sample API Server using the current Aggregator [Conformance]
```

<details>
<summary><b>environment</b></summary>

```
GO111MODULE=on go get "sigs.k8s.io/kind@v0.7.0"
kind version
kind v0.7.0 go1.14.2 linux/amd64

sonobuoy version
Sonobuoy Version: v0.18.0
MinimumKubeVersion: 1.16.0
MaximumKubeVersion: 1.18.99

docker version
Client: Docker Engine - Community
 Version:           19.03.6
 API version:       1.40
 Go version:        go1.12.16
 Git commit:        369ce74a3c
 Built:             Thu Feb 13 01:27:49 2020
 OS/Arch:           linux/amd64
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          19.03.6
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.16
  Git commit:       369ce74a3c
  Built:            Thu Feb 13 01:26:23 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.12
  GitCommit:        35bd7a5f69c13e1563af8a93431411cd9ecf5021
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683

cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=19.04
DISTRIB_CODENAME=disco
DISTRIB_DESCRIPTION="Pop!_OS 19.04"

uname -srvio
Linux 5.0.0-21-generic #22+system76-Ubuntu SMP Tue Jul 16 19:57:52 UTC 2019 x86_64 GNU/Linux
```

</details>

**Comment by stealthybox** [comment]
`Should be able to support the 1.10 Sample API Server using the current Aggregator [Conformance]` passed in the [1.17 kubeadm results](https://github.com/cncf/k8s-conformance/pull/932).
I did not experience it flaking at all.

**Comment by BenTheElder** [comment]
thanks!

FWIW the conformance tests are slightly flaky pretty much everywhere, e.g. even this has flaked https://testgrid.k8s.io/conformance-gce#Conformance%20-%20GCE%20-%201.17&width=5&sort-by-flakiness=

... but I wouldn't expect it to consistently fail. it seems we have a hole in our CI dashboard here as well. not sure why the test-infra config forker didn't the 1.17, 1.18 entries for kind ...

**Comment by BenTheElder** [comment]
this was the 1.18 change https://github.com/kubernetes/kubernetes/pull/84735
I'd be inclined to try with the latest 1.17.X next.

@liggitt do you know of any issues with this test on < 1.18?

**Comment by liggitt** [comment]
no... that test hasn't really changed in many releases... there was no change between 1.16 and 1.17...

the change in 1.18 was to a 1.17 image that would only use v1 authn/authz delegation APIs as part of the no-beta conformance effort

```sh
git diff upstream/release-1.16 upstream/release-1.17 test/e2e/apimachinery/aggregator.go
```

```diff
diff --git a/test/e2e/apimachinery/aggregator.go b/test/e2e/apimachinery/aggregator.go
index 8e9a8faf9fd..fefe5954dff 100644
--- a/test/e2e/apimachinery/aggregator.go
+++ b/test/e2e/apimachinery/aggregator.go
@@ -41,7 +41,6 @@ import (
 	rbacv1helpers "k8s.io/kubernetes/pkg/apis/rbac/v1"
 	"k8s.io/kubernetes/test/e2e/framework"
 	e2edeploy "k8s.io/kubernetes/test/e2e/framework/deployment"
-	e2elog "k8s.io/kubernetes/test/e2e/framework/log"
 	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
 	imageutils "k8s.io/kubernetes/test/utils/image"
 	samplev1alpha1 "k8s.io/sample-apiserver/pkg/apis/wardle/v1alpha1"
@@ -81,11 +80,11 @@ var _ = SIGDescribe("Aggregator", func() {
 		if aggrclient == nil {
 			config, err := framework.LoadConfig()
 			if err != nil {
-				e2elog.Failf("could not load config: %v", err)
+				framework.Failf("could not load config: %v", err)
 			}
 			aggrclient, err = aggregatorclient.NewForConfig(config)
 			if err != nil {
-				e2elog.Failf("could not create aggregator client: %v", err)
+				framework.Failf("could not create aggregator client: %v", err)
 			}
 		}
 	})
@@ -375,16 +374,16 @@ func TestSampleAPIServer(f *framework.Framework, aggrclient *aggregatorclient.Cl
 	}, "Waited %s for the sample-apiserver to be ready to handle requests.")
 	if err != nil {
 		currentAPIServiceJSON, _ := json.Marshal(currentAPIService)
-		e2elog.Logf("current APIService: %s", string(currentAPIServiceJSON))
+		framework.Logf("current APIService: %s", string(currentAPIServiceJSON))
 
 		currentPodsJSON, _ := json.Marshal(currentPods)
-		e2elog.Logf("current pods: %s", string(currentPodsJSON))
+		framework.Logf("current pods: %s", string(currentPodsJSON))
 
 		if currentPods != nil {
 			for _, pod := range currentPods.Items {
 				for _, container := range pod.Spec.Containers {
 					logs, err := e2epod.GetPodLogs(client, namespace, pod.Name, container.Name)
-					e2elog.Logf("logs of %s/%s (error: %v): %s", pod.Name, container.Name, err, logs)
+					framework.Logf("logs of %s/%s (error: %v): %s", pod.Name, container.Name, err, logs)
 				}
 			}
 		}
@@ -402,7 +401,7 @@ func TestSampleAPIServer(f *framework.Framework, aggrclient *aggregatorclient.Cl
 	var statusCode int
 	result.StatusCode(&statusCode)
 	if statusCode != 201 {
-		e2elog.Failf("Flunders client creation response was status %d, not 201", statusCode)
+		framework.Failf("Flunders client creation response was status %d, not 201", statusCode)
 	}
 
 	pods, err := client.CoreV1().Pods(namespace).List(metav1.ListOptions{})
@@ -416,7 +415,7 @@ func TestSampleAPIServer(f *framework.Framework, aggrclient *aggregatorclient.Cl
 	err = json.Unmarshal(contents, &flundersList)
 	validateErrorWithDebugInfo(f, err, pods, "Error in unmarshalling %T response from server %s", contents, "/apis/wardle.k8s.io/v1alpha1")
 	if len(flundersList.Items) != 1 {
-		e2elog.Failf("failed to get back the correct flunders list %v", flundersList)
+		framework.Failf("failed to get back the correct flunders list %v", flundersList)
 	}
 
 	// kubectl delete flunder test-flunder -v 9
@@ -431,7 +430,7 @@ func TestSampleAPIServer(f *framework.Framework, aggrclient *aggregatorclient.Cl
 	err = json.Unmarshal(contents, &flundersList)
 	validateErrorWithDebugInfo(f, err, pods, "Error in unmarshalling %T response from server %s", contents, "/apis/wardle.k8s.io/v1alpha1")
 	if len(flundersList.Items) != 0 {
-		e2elog.Failf("failed to get back the correct deleted flunders list %v", flundersList)
+		framework.Failf("failed to get back the correct deleted flunders list %v", flundersList)
 	}
 
 	flunderName = generateFlunderName("dynamic-flunder")
@@ -443,7 +442,7 @@ func TestSampleAPIServer(f *framework.Framework, aggrclient *aggregatorclient.Cl
 	gvr := schema.GroupVersionResource{Group: "wardle.k8s.io", Version: "v1alpha1", Resource: "flunders"}
 	_, ok := groupVersionResources[gvr]
 	if !ok {
-		e2elog.Failf("could not find group version resource for dynamic client and wardle/flunders (discovery error: %v, discovery results: %#v)", discoveryErr, groupVersionResources)
+		framework.Failf("could not find group version resource for dynamic client and wardle/flunders (discovery error: %v, discovery results: %#v)", discoveryErr, groupVersionResources)
 	}
 	dynamicClient := f.DynamicClient.Resource(gvr).Namespace(namespace)
 
@@ -469,7 +468,7 @@ func TestSampleAPIServer(f *framework.Framework, aggrclient *aggregatorclient.Cl
 	unstructuredList, err := dynamicClient.List(metav1.ListOptions{})
 	framework.ExpectNoError(err, "listing flunders using dynamic client")
 	if len(unstructuredList.Items) != 1 {
-		e2elog.Failf("failed to get back the correct flunders list %v from the dynamic client", unstructuredList)
+		framework.Failf("failed to get back the correct flunders list %v from the dynamic client", unstructuredList)
 	}
 
 	// kubectl delete flunder test-flunder
@@ -480,19 +479,19 @@ func TestSampleAPIServer(f *framework.Framework, aggrclient *aggregatorclient.Cl
 	unstructuredList, err = dynamicClient.List(metav1.ListOptions{})
 	framework.ExpectNoError(err, "listing flunders using dynamic client")
 	if len(unstructuredList.Items) != 0 {
-		e2elog.Failf("failed to get back the correct deleted flunders list %v from the dynamic client", unstructuredList)
+		framework.Failf("failed to get back the correct deleted flunders list %v from the dynamic client", unstructuredList)
 	}
 
 	cleanTest(client, aggrclient, namespace)
 }
 
 // pollTimed will call Poll but time how long Poll actually took.
-// It will then e2elog.Logf the msg with the duration of the Poll.
+// It will then framework.Logf the msg with the duration of the Poll.
 // It is assumed that msg will contain one %s for the elapsed time.
 func pollTimed(interval, timeout time.Duration, condition wait.ConditionFunc, msg string) error {
 	defer func(start time.Time, msg string) {
 		elapsed := time.Since(start)
-		e2elog.Logf(msg, elapsed)
+		framework.Logf(msg, elapsed)
 	}(time.Now(), msg)
 	return wait.Poll(interval, timeout, condition)
 }
@@ -513,7 +512,7 @@ func validateErrorWithDebugInfo(f *framework.Framework, err error, pods *v1.PodL
 			msg += fmt.Sprintf("\nOriginal pods in %s:\n%v", namespace, pods)
 		}
 
-		e2elog.Failf(msg)
+		framework.Failf(msg)
 	}
 }
 
```

**Comment by liggitt** [comment]
what is the failure you're seeing on 1.17?

**Comment by neolit123** [comment]
we have this test [disabled](https://github.com/neolit123/kubeadm/blob/bf4c4d2465f9d9988e39bb7988618bff9d084449/kinder/cmd/kinder/test/e2e/e2e.go#L79) in kubeadm/kinder testing.
then we faced the same test failure when we enabled e2e tests for kubeadm/windows so we [disabled it there too](https://github.com/kubernetes-sigs/sig-windows-tools/blob/c9e6961aa8bead79d9303c276f58086d72aa355c/kubeadm/hack/e2e-cluster-gcp.sh#L150)
but the test appears to be fixed in 1.18.

i tried understanding why it is happening but didn't have enough time.
no recent logs, but it was something related to "200 status" and "error bodies".

the failure was always reproducible with kubeadm, so not sure how it's passing for kind conformance here: https://testgrid.k8s.io/conformance-gce#Conformance%20-%20GCE%20-%201.17&width=5&sort-by-flakiness=

i can try running it again later with `--ginkgo.focus=Aggregator` to obtain the logs...

a kubeadm api-server pod looks like:

<details>

```
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: null
  labels:
    component: kube-apiserver
    tier: control-plane
  name: kube-apiserver
  namespace: kube-system
spec:
  containers:
  - command:
    - kube-apiserver
    - --advertise-address=192.168.0.102
    - --allow-privileged=true
    - --authorization-mode=Node,RBAC
    - --client-ca-file=/etc/kubernetes/pki/ca.crt
    - --enable-admission-plugins=NodeRestriction
    - --enable-bootstrap-token-auth=true
    - --etcd-cafile=/etc/kubernetes/pki/etcd/ca.crt
    - --etcd-certfile=/etc/kubernetes/pki/apiserver-etcd-client.crt
    - --etcd-keyfile=/etc/kubernetes/pki/apiserver-etcd-client.key
    - --etcd-servers=https://127.0.0.1:2379
    - --insecure-port=0
    - --kubelet-client-certificate=/etc/kubernetes/pki/apiserver-kubelet-client.crt
    - --kubelet-client-key=/etc/kubernetes/pki/apiserver-kubelet-client.key
    - --kubelet-preferred-address-types=InternalIP,ExternalIP,Hostname
    - --proxy-client-cert-file=/etc/kubernetes/pki/front-proxy-client.crt
    - --proxy-client-key-file=/etc/kubernetes/pki/front-proxy-client.key
    - --requestheader-allowed-names=front-proxy-client
    - --requestheader-client-ca-file=/etc/kubernetes/pki/front-proxy-ca.crt
    - --requestheader-extra-headers-prefix=X-Remote-Extra-
    - --requestheader-group-headers=X-Remote-Group
    - --requestheader-username-headers=X-Remote-User
    - --secure-port=6443
    - --service-account-key-file=/etc/kubernetes/pki/sa.pub
    - --service-cluster-ip-range=10.96.0.0/12
    - --tls-cert-file=/etc/kubernetes/pki/apiserver.crt
    - --tls-private-key-file=/etc/kubernetes/pki/apiserver.key
    image: k8s.gcr.io/kube-apiserver:v1.14.2
    imagePullPolicy: IfNotPresent
    livenessProbe:
      failureThreshold: 8
      httpGet:
        host: 192.168.0.102
        path: /healthz
        port: 6443
        scheme: HTTPS
      initialDelaySeconds: 15
      timeoutSeconds: 15
    name: kube-apiserver
    resources:
      requests:
        cpu: 250m
    volumeMounts:
    - mountPath: /etc/ssl/certs
      name: ca-certs
      readOnly: true
    - mountPath: /etc/ca-certificates
      name: etc-ca-certificates
      readOnly: true
    - mountPath: /etc/pki
      name: etc-pki
      readOnly: true
    - mountPath: /etc/kubernetes/pki
      name: k8s-certs
      readOnly: true
    - mountPath: /usr/local/share/ca-certificates
      name: usr-local-share-ca-certificates
      readOnly: true
    - mountPath: /usr/share/ca-certificates
      name: usr-share-ca-certificates
      readOnly: true
  hostNetwork: true
  priorityClassName: system-cluster-critical
  volumes:
  - hostPath:
      path: /etc/ssl/certs
      type: DirectoryOrCreate
    name: ca-certs
  - hostPath:
      path: /etc/ca-certificates
      type: DirectoryOrCreate
    name: etc-ca-certificates
  - hostPath:
      path: /etc/pki
      type: DirectoryOrCreate
    name: etc-pki
  - hostPath:
      path: /etc/kubernetes/pki
      type: DirectoryOrCreate
    name: k8s-certs
  - hostPath:
      path: /usr/local/share/ca-certificates
      type: DirectoryOrCreate
    name: usr-local-share-ca-certificates
  - hostPath:
      path: /usr/share/ca-certificates
      type: DirectoryOrCreate
    name: usr-share-ca-certificates
status: {}
```

</details>

**Comment by neolit123** [comment]
TL;DR i couldn't get the test to fail in normal circumstances.

```bash
# setup kind cluster
$ kind version
kind v0.7.0 go1.13.9 linux/amd64
$ cat config.yaml
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
- role: control-plane
- role: worker
- role: worker
$ kind create cluster --config config.yaml --image kindest/node:v1.17.0
$ kubectl get no
NAME                 STATUS   ROLES    AGE   VERSION
kind-control-plane   Ready    master   21m   v1.17.0
kind-worker          Ready    <none>   20m   v1.17.0
kind-worker2         Ready    <none>   20m   v1.17.0
$ export KUBECONFIG=~/.kube/config

# build e2e.test from 1.17
$ cd kubernetes/kubernetes
$ git pull
$ git checkout v1.17.0
$ make all WHAT=test/e2e/e2e.test

# serial run
$ for i in {1..100}; do ./_output/local/go/bin/e2e.test --ginkgo.focus=Aggregator 2>&1 | grep "Passed"; done
# all 100 runs passed...

# parallel run
# NOTE: these parallel runs are NOT a real world test...as we never run the same test in parallel.
#
$ for i in {1..2}; do (./_output/local/go/bin/e2e.test --ginkgo.focus=Aggregator &> log$i.txt & ); done
#
# https://pastebin.com/GaGLuCdM
#
# one issue is the test reserves the "sample-apiserver-reader" clusterrole without random suffix in the name.
# "clusterroles.rbac.authorization.k8s.io "sample-apiserver-reader" already exists"
there is another error related to the fluder in the test:
#  "flunders.wardle.k8s.io "rest-flunder-2127593694" is forbidden: not yet ready to handle request"
```

**Comment by stealthybox** [comment]
1.16.4 also fails on my machine using `run.sh` with the same environment.
```shell
e2e_focus="Sample API Server" ./run.sh 1.16.4
```
I also modified the script to support 1.16.1 and it also failed.

This gave me the thought that maybe this test wasn't part of the `--mode=certified-conformance` test suite added in 1.16. The test was included, and it passed in Ben's submission: [cncf/k8s-conformance/v1.16/kind/e2e.log (raw)](https://raw.githubusercontent.com/cncf/k8s-conformance/2a5c5b1fc46d6bb732e84f15491ae738d0d03a24/v1.16/kind/e2e.log)

[kind#906](https://github.com/cncf/k8s-conformance/pull/906/files#diff-15f13414bba482dc30c2d221b9f3dcf0R3) was apparently submitted using kind v0.7.0 and an older version of sonobuoy.

I can't get this test to pass using kind v0.7.0 and sonobuoy 0.18.0 on my machine.
Can any of you replicate the failure with these versions using `run.sh` or is this just broken for me?

**Comment by BenTheElder** [comment]
> the failure was always reproducible with kubeadm, so not sure how it's passing for kind conformance here: https://testgrid.k8s.io/conformance-gce#Conformance%20-%20GCE%20-%201.17&width=5&sort-by-flakiness=

that is GCE, not kind.

I'll try to look at replicating this soon, I'm triaging my backlog atm and this is reasonably high ...

**Comment by BenTheElder** [comment]
these are mostly done, we'll of course need to continue doing this in the future.

**Comment by BenTheElder** [comment]
we're going to need to do this again soon 🙃 
tempted to automate this ...
